journalpump/journalpump.py

```diff
@@ -233,21 +235,29 @@ def initialize_senders(self):
             if not isinstance(extra_field_values, dict):
                 self.log.warning("extra_field_values: %r not a dictionary object, ignoring", extra_field_values)
                 extra_field_values = {}
+            try:
+                sender = sender_class(
+                    config=sender_config,
+                    field_filter=field_filter,
+                    extra_field_values=extra_field_values,
+                    msg_buffer_max_length=self.msg_buffer_max_length,
+                    name=sender_name,
+                    reader=self,
+                    stats=self.stats,
+                    tags=self.make_tags(),
+                )
+            except SenderError:
+                # If sender init fails, log exception, don't start() the sender
+                # and don't add it to self.senders dict. A metric about senders that
+                # failed to start is sent at the end of this method
+                self.log.exception("Sender %r failed to initialize", sender_name)
```

Comment:
I think it would be good to retry at least few times before deciding the sender has failed. This might be a bit sender specific but at least the AWS sender uses `LogSender._backoff` to throttle retries and this is used if there are errors while sending the log messages upstream. I think the sending of the log messages is retried indefinitely (sleeping max 30 seconds between retries), but here it could maybe retry a few times and then decide the sender has failed.

Reply:
Ah, sorry. It is raising `SenderError` only if authentication fails. So maybe no reason to retry that. Maybe rename the `SenderError` to something less generic like `SenderInitializationError`?

Reply:
If it gets pass the `LogSender.start` method the retrying will actually happen when sending messages.

Reply:
The authentication failure happens in `AWSCloudWatchSender(LogSender).__init__()` so `LogSender.run()` has not been called yet. Adding a few retries with `LogSender._backoff()` would still be easy to do in `AWSCloudWatchSender._init_logs()`. AWS does return transient 503 errors occasionally, so retrying would handle this case nicely as well.

I'll rename `SenderError` to `SenderInitializationError` to avoid confusion.